### PR TITLE
Feature/OIDC

### PR DESCRIPTION
**Pull request by temi**

### CODE REVIEW

**Comment by temi** [comment]
https://github.com/AtlasOfLivingAustralia/biocollect/issues/1414

File: grails-app/conf/application.yml [hunk 1]
```diff
@@ -19,21 +19,39 @@ app:
...
+    enabled: true
+    discoveryUri: "${auth.baseURL}/cas/oidc/.well-known"
+    clientId: "auth-dev-oidc-client-id"
+    secret: "u85789gnghdfb765"
+    scope: "openid,profile,email,ala,roles,user_defined"
+    allowUnsignedIdTokens: true
+  jwt:
+    enabled: true
+    discoveryUri: "${auth.baseURL}/cas/oidc/.well-known"
+    requiredClaims: ["sub", "iat", "exp", "jti", "client_id"]
```
Comment (yasima-csiro) [inline grails-app/conf/application.yml:54]: No need to set up requiredClaims now.

File: grails-app/conf/application.yml [hunk 2]
```diff
@@ -19,21 +19,39 @@ app:
...
-#      casServerName: "https://auth.ala.org.au"
-#      casServerLoginUrl: "https://auth.ala.org.au/cas/login"
-#      casServerUrlPrefix: "https://auth.ala.org.au/cas"
-#      logoutUrl: "https://auth.ala.org.au/cas/logout"
-#      loginUrl: "https://auth.ala.org.au/cas/login"
+  cas:
+    enabled: false
+    uriExclusionFilterPattern: ['/assets/.*','/ajax/(saveBookmarkLocation|getBookmarkLocations)','/uploads/.*']
+    authenticateOnlyIfCookieFilterPattern: "/*"
+    uriFilterPattern: ['/user/.*','/site/(?!(index|list|elasticsearch|getImages|getPoiImages|ajaxUpdate|checkSiteName|ajaxList|checkPointInsideProjectAreaAndAddress)).*','/project/(?!(index|search|citizenScience|listRecordImages|projectSummaryReportCallback)).*','/activity/(?!index).*','/output/(?!index).*','/image/(?!index).*','/admin/.*','/proxy/speciesListPost','/proxy/documentUpdate','/proxy/deleteDocument','/home/advanced','/organisation/(?!index).*','/organisation/(?!list).*','/bioActivity/create/.*','/bioActivity/edit/.*','/bioActivity/list','/sight/.*']
```
Comment (yasima-csiro) [inline grails-app/conf/application.yml:33]: uriFilterPattern does not support regex values. only supports Java Servlet Filter paths.

https://github.com/AtlasOfLivingAustralia/ala-auth-plugin#setup-cas-authentication-for-your-app

File: grails-app/views/layouts/mobile.gsp [hunk 3]
```diff
@@ -59,11 +60,16 @@
             withCredentials: true
         },
         beforeSend: function (xhr) {
-            xhr.setRequestHeader('authKey', "${authKey}");
-            xhr.setRequestHeader('userName', "${userName}");
+            <g:if test="${authorization}">
+                xhr.setRequestHeader('Authorization', "${authorization}");
+            </g:if>
+            <g:elseif test="${grailsApplication.config.getProperty("mobile.authKeyEnabled", Boolean) && authKey && userName}">
+                xhr.setRequestHeader('authKey', "${authKey}");
```
Comment (yasima-csiro) [inline grails-app/views/layouts/mobile.gsp:67]: I am not sure this will work with Cognito

File: build.gradle [hunk 4]
```diff
@@ -113,12 +114,14 @@ dependencies {
...
+        exclude group: 'org.grails.plugins', module: 'ala-auth'
     }
 
 
     implementation "org.grails.plugins:ala-admin-plugin:2.2"
-    implementation "org.grails.plugins:ala-auth:3.2.3"
-    runtimeOnly "org.grails.plugins:ala-bootstrap3:3.2.3"
+    implementation "org.grails.plugins:ala-auth:5.0.0-SNAPSHOT"
+    implementation "org.grails.plugins:ala-ws-security-plugin:4.0.0-SNAPSHOT"
+    runtimeOnly "org.grails.plugins:ala-bootstrap3:4.0.1-SNAPSHOT"
```
Comment (yasima-csiro) [inline build.gradle:124]: New plugins available:
ala-auth:5.0.2-SNAPSHOT
ala-ws-security-plugin:4.0.0
ala-bootstrap3:4.0.0

File: grails-app/views/layouts/mobile.gsp (hunk 3 above)
Comment (temi) [reply grails-app/views/layouts/mobile.gsp:67]: API will be disabled before moving to cognito.

File: build.gradle [hunk 5]
```diff
@@ -123,8 +125,12 @@ dependencies {
...
 
     implementation "org.grails.plugins:ala-admin-plugin:2.2"
-    implementation "org.grails.plugins:ala-auth:3.2.3"
-    runtimeOnly "org.grails.plugins:ala-bootstrap3:3.2.3"
+    implementation ("org.grails.plugins:ala-auth:5.1.1")
+    implementation "org.grails.plugins:ala-ws-security-plugin:4.1.2"
+    runtimeOnly "org.grails.plugins:ala-bootstrap3:4.0.0"
+
+    //    swagger API
+    implementation 'au.org.ala.plugins:openapi:1.0.0'
```
Comment (yasima-csiro) [inline build.gradle:133]: Shall we use latest plugin - 1.1.0?

File: grails-app/conf/application.yml [hunk 6]
```diff
@@ -19,21 +19,38 @@ app:
...
+    casServerUrlPrefix: "${auth.baseURL}/cas"
+    logoutUrl: "${security.cas.casServerUrlPrefix}/logout"
+    loginUrl: "${security.cas.casServerUrlPrefix}/login"
+
+  oidc:
+    enabled: true
+    discoveryUri: "${auth.baseURL}/cas/oidc/.well-known"
+    clientId: "changeMe"
+    secret: "changeMe"
+    scope: "openid,profile,email,ala,roles,user_defined"
```
Comment (yasima-csiro) [inline grails-app/conf/application.yml:48]: Don't need "user_defined" scope

File: grails-app/conf/application.yml [hunk 7]
```diff
@@ -19,21 +19,38 @@ app:
...
+    allowUnsignedIdTokens: true
+  jwt:
+    enabled: true
+    discoveryUri: "${auth.baseURL}/cas/oidc/.well-known"
+    requiredClaims: ["sub", "iat", "exp", "jti", "client_id"]
```
Comment (yasima-csiro) [inline grails-app/conf/application.yml:54]: No need to set up requiredClaims.

File: grails-app/controllers/au/org/ala/biocollect/BioActivityController.groovy [hunk 8]
```diff
@@ -46,6 +64,55 @@ class BioActivityController {
...
+            ],
+            responses = [
+                    @ApiResponse(
+                            responseCode = "200",
+                            content = @Content(
+                                    mediaType = "application/json",
+                                    schema = @Schema(
+                                            implementation = ActivitySaveResponse.class
+                                    )
+                            )
```
Comment (yasima-csiro) [inline grails-app/controllers/au/org/ala/biocollect/BioActivityController.groovy:101]: Can you please add CORS headers section for all 200 responses?

https://confluence.csiro.au/display/ALASD/OIDC+Migration+on+ALA+Applications

File: grails-app/controllers/au/org/ala/biocollect/merit/ImageController.groovy [hunk 9]
```diff
@@ -134,6 +155,56 @@ class ImageController {
...
+                        responseCode = "200",
+                        content = @Content(
+                                mediaType = "application/json",
+                                schema =  @Schema(
+                                        oneOf = [
+                                                UploadResponse.class,
+                                                FileUploadErrorResponse.class
+                                        ]
+                                )
+                        )
```
Comment (yasima-csiro) [inline grails-app/controllers/au/org/ala/biocollect/merit/ImageController.groovy:201]: Can you please add CORS headers section as per https://confluence.csiro.au/display/ALASD/OIDC+Migration+on+ALA+Applications?

File: grails-app/controllers/au/org/ala/biocollect/merit/OutputController.groovy [hunk 10]
```diff
@@ -147,6 +158,34 @@ class OutputController {
...
+            summary = "Get Output Species Identifier. It is an UUID assigned to each species on a form.",
+            responses = [
+                    @ApiResponse(
+                            responseCode = "200",
+                            content = @Content(
+                                    mediaType = "application/json",
+                                    schema = @Schema(
+                                            implementation = GetOutputSpeciesIdentifierResponse.class
+                                    )
+                            )
```
Comment (yasima-csiro) [inline grails-app/controllers/au/org/ala/biocollect/merit/OutputController.groovy:174]: Please add CORS headers section - https://confluence.csiro.au/display/ALASD/OIDC+Migration+on+ALA+Applications

File: grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy [hunk 11]
```diff
@@ -48,7 +67,64 @@ class ProjectController {
...
+                            responseCode = "200",
+                            content = @Content(
+                                    mediaType = "application/json",
+                                    array = @ArraySchema(
+                                            schema = @Schema(
+                                                    implementation = Map.class
+                                            )
+                                    )
+                            )
+
```
Comment (yasima-csiro) [inline grails-app/controllers/au/org/ala/biocollect/merit/ProjectController.groovy:103]: Please add CORS headers section - https://confluence.csiro.au/display/ALASD/OIDC+Migration+on+ALA+Applications

**Review by yasima-csiro — CHANGES_REQUESTED** [review]
Can you please address review comments?
Thanks.

**Comment by temi** [comment]
@yasima-csiro those comments are addressed now.

**Review by yasima-csiro — APPROVED** [review]

**Pull request merged**